### ISSUE TITLE
chatDidReceivePresenceWithStatus not being called after sendPresenceWithStatus is called.

### ISSUE THREAD

**Issue by astanton** [QuickBlox/quickblox-ios-sdk#402]
First I'll say I am brand new to QB and I'm currently going through the video chat tutorial for iOS.  So I may be doing something wrong here from a conceptual point of view.

I have 2 devices connected, and I have added a button action to the sample app that will call `sendPresenceWithStatus' and I pass a status of a string.

I then also have that class as a `QBChatDelegate` and I am implementing `chatDidReceivePresenceWithStatus`.  However, it's not working as I expected.

As user A I am calling 'sendPresenceWithStatus' and I'm expecting user B to have the 'chatDidReceivePresenceWithStatus' delegate method to be called.  However, that method is being called user A's client but not user B's.  

Am I just expecting this to work differently than it is?  This chat only has these 2 users connected to it.

What I'm trying to do in general is just send a teeny tiny bit of information across to the other user (or both users at the same time too) when certain actions happen.  I was trying to do that by just sending a status of a string that I needed.

If this isn't working as I am expecting, is there a method that I can send signals between 2 users that are connected just to send booleans or strings?  It's very little data.  

I thought I read that internally the video chat is using signals via the QBChat api, and I am kind of trying to do that in a custom way.

**Comment by VitaliyGuru** [comment]
Hi @astanton ,
Presences work only for users which are subscribed to each other. 
Please look at [this explanation of presences and subscribes](http://xmpp.org/rfcs/rfc3921.html#int)

If you want to send presence to users you should  request a subscription to this users, and they should accept this request by following methods from QBChat:
`- (void)addUserToContactListRequest:(NSUInteger)userID completion:(QB_NULLABLE QBChatCompletionBlock)completion;` - request to subscribe

`- (void)confirmAddContactRequest:(NSUInteger)userID completion:(QB_NULLABLE QBChatCompletionBlock)completion;` - confirm user's subscribe request

Otherwise if you only want to send custom data you can use this method from QBChatDialog:
`- (void)sendMessage:(QB_NONNULL QBChatMessage *)message completionBlock:(QB_NULLABLE_S QBChatCompletionBlock)completion ` 

`QBChatMessage` has property `customParameters`, so you can set and send your custom data to other user by adding custom parameters to this mutable dictionary.

Of course, you should write your own logic for parsing and getting this parameters by opponent user.

Also you can use method :
`- (void)sendSystemMessage:(QB_NONNULL QBChatMessage *)message completion:(QB_NULLABLE QBChatCompletionBlock)completion;`

for sending messages to specific users.

For more further explanation and examples please refer to our [services](https://github.com/QuickBlox/q-municate-services-ios#qmchatservice). They are straightforward and easy to implement. 

There you can find methods for sending system and text messages with additional handling.

Feel free to ask any questions.

**Comment by astanton** [comment]
Thanks for the long detailed explanation.  That helps me understand a bit more what the presence methods are being used for.

As for the `sendSystemMessage` method, I was under the impression that those were sent to all users who are currently logged in, based off of the wording at `http://quickblox.com/developers/SimpleSample-chat_users-ios#System_notifications` where it says they will only be delivered to online users.  But I see now it takes a parameter to a userID, and I'm now assuming that it sends a message to that user and will only get to them if they are online and NOT to all users online?

Also, can you send system messages to more than 1 user at a time, so that I could have user A send a system message to user A and user B at the same time?

As for the method `- (void)sendMessage:(QB_NONNULL QBChatMessage *)message completionBlock:(QB_NULLABLE_S QBChatCompletionBlock)completion` is that something that is stored on the server?  

My use case is to basically have 2 users in a video chat, user A and user B, and I want user A (or B) to perform an action (such as push a button) and then be able to notify both users A and B at the same time that this action was performed.  I was hoping there was a way I could send some sort of signal/message to both users at the same time so that they would receive them at the same time, to keep things in sync.  I don't need to keep any kind of history of these actions either, it's just a one time thing to let the clients know what to do while they are in a chat.

Does that make sense?  If so, which way would you recommend doing that via these different methods?  

I guess in theory I could use the presence methods with status, but that would require having A and B added to contact list at beginning of chat, then using the `status` string as the "action" parameterj, and after the video chat is done I'd have to remove them from the contact list since their session is over.  

Would that be an alright way to achieve what I'm going for, or is there another way you would recommend doing this?

Thanks.

**Comment by VitaliyGuru** [comment]
> As for the method - (void)sendMessage:(QB_NONNULL QBChatMessage *)message completionBlock:(QB_NULLABLE_S QBChatCompletionBlock)completion is that something that is stored on the server?

Yes.

We recommend you to use system messages. They are sent only for specific users and NOT to all users online.

In [this](https://github.com/QuickBlox/q-municate-services-ios/blob/master/QMChatService/QMChatService/QMChatService.m#L1307-1336) method you can find out how to send system message to specific users. QMChatService is a part of our [QMServices](https://github.com/QuickBlox/q-municate-services-ios)

**Comment by astanton** [comment]
Okay thanks again.  

A couple more questions.

1. Is there anyway to send a system message to multiple users at the same time?  Could I have user A send a system message to A and B in the same call?  From the looks of it, there appears to only be 1 recipient userID.  If not, would sending 2 system messages (one to A and one to B) in succession be a way I could do this?  Would both of those messages be received at the same time?

The reason I ask is because I want user A to perform an action, then i want both user A and user B to react to that action at the same time (or within milliseconds of eachother).  

2.  The systemMessage calls you show in there go along with a dialog.  Couldn't I just use the `[[QBChat instance] sendSystemMessage:] method instead of using the service call that also uses a dialog?  And then in the delegate I could just handle it in the `chatDidReceiveSystemMessage:` method?  

I guess I am not exactly sure what extra benefit the dialog would give me, but it could be that I'm misunderstanding exactly what is necessary for this.

That is all for now thank you again for the help, you are giving me a direction to go with this!

**Comment by VitaliyGuru** [comment]
> Is there anyway to send a system message to multiple users at the same time? Could I have user A send a system message to A and B in the same call? From the looks of it, there appears to only be 1 recipient userID. If not, would sending 2 system messages (one to A and one to B) in succession be a way I could do this? Would both of those messages be received at the same time?

 There will be a little delay.

> The systemMessage calls you show in there go along with a dialog. Couldn't I just use the [[QBChat instance] sendSystemMessage:] method instead of using the service call that also uses a dialog? And then in the delegate I could just handle it in thechatDidReceiveSystemMessage:` method?

Of course,  I just showed you an example how to use it. In services we use this method with dialog, cause from the dialog we get some parameters and set them to the custom parameters of message.

**Comment by astanton** [comment]
Okay thanks a ton for all of the information.  It helped clear up quite a bit and I have a direction I plan on trying out this evening.

**Comment by VitaliyGuru** [comment]
@astanton , The delay depends on server. So I can't give you one hundred percent sure answer.
You can test it and decide is this exactly what you're looking for or not.

**Comment by astanton** [comment]
Thanks for getting back to me.

So in theory though, if I have user A and user B on each other's contact list, and then I had user A call `sendPresenceWithStatus`, both user A and user B would have the `chatDidReceivePresenceWithStatus` method called from that presence message being sent out, correct?   If that is the case, I can do some testing and see what the timestamp would be when they both receive them and see how close they are.

I will also try sending out 2 system messages in succession, one to user A and one to user B, and do a timestamp comparison on when they both receive them in the delegate method.

There are some instances where I want user A to send a message that only user B receives, and then there are other instances where I want user A to send a message that both user A and B receive, as close to the same time as possible.  Perhaps my solution is to have a mix of presence messages with statuses that go to both users, and then system messages that go from one user to the other only.

**Comment by VitaliyGuru** [comment]
> So in theory though, if I have user A and user B on each other's contact list, and then I had user A call sendPresenceWithStatus, both user A and user B would have the chatDidReceivePresenceWithStatus method called from that presence message being sent out, correct?

Yes, you are right.

When you send presence you can send only one parameter `status`.
As for system message you can send multiple parameters via `customParameters` of the message.

**Comment by astanton** [comment]
Okay gotcha.  

The "messages" I need to send are very short so the string with the presence method will work just fine.  I can also get the user ID of the sender as that is one of the parameters in the callback too, so that works.

However I will test out the system messages first and if the 2 messages are received milliseconds apart then I'd rather use that, just so I can deal only with one type of message, as well as not having to setup/cleanup the contact lists for the presence messages. 

But I will give it a go tonight and see what I can come up with.

Thanks again for all of your help.

**Comment by VitaliyGuru** [comment]
@astanton , You're welcome.

**Comment by VitaliyGuru** [comment]
Hi ,@astanton  
Have you chosen the solution for your needs?

**Comment by astanton** [comment]
@VitaliyGuru I have actually.  I am going to be using system messages.

I did some testing last night.  I ended up having user A send a system message to user B and then to user A (in that order).  I outputted the current time, with milliseconds, to the console when each message was sent.  

Then in the system received delegate method I simply had it output that the message was received and at what time.

The time between sending and receiving was < 100 milliseconds, and I believe it was even < 50 milliseconds.  And the difference between when user B and user A received the message was negligible and within 50 milliseconds of eachother as well, and I believe even less than that.

I did not even bother with the presence messages because this does exactly what I needed to do by simply being able to use the `message.text` attribute, along with the `senderID` and `recipientID` so this is the plan I am going to approach this with.

Once I get the face detection issue resolved (in the other thread) then I'm going to start working on converting our app to QuickBlox from OpenTok since I will then know that the 2 main features I need will be working in the sample video chat app.
